go/parquet/file/page_reader.go

```diff
@@ -611,8 +610,6 @@ func (p *serializedPageReader) Next() bool {
 			// we don't know this page type, we're allowed to skip non-data pages
 			continue
 		}
-
-		p.buf = memory.NewResizableBuffer(p.mem)
```

Comment:
Since a new `memory.Buffer` was created at the end here, I opted to create (and release) a new buffer at the start of the function instead.

Alternative fix would be:
* remove this, so that the buffer is reused across calls to `serializedPageReader.Next` (is this safe?)
* add a `Release` method to the `PageReader` interface
* have the `primitiveRecordReader` release the `PageReader`

Reply:
Ah I see what was happening. I passed the raw bytes as a *new* buffer to the page itself. So technically even though the accounting in the memory allocator is now correct, if the allocator isn't a Go allocator this will actually end up using the buffer after it's been freed. if you look at line 596, it calls `memory.NewBufferBytes(data)` to pass the buffer to the datapage itself. `NewBufferBytes` constructs a buffer which wraps around a byte slice, but it doesn't attempt to *free* that slice when it is garbage collected and assumes that the Go garbage collector will handle it. 

This is all fine for a byteslice that is allocated by Go itself, but since this is a buffer coming from the resizable buffer allocated using the `memory.Allocator` it's possible that this memory wasn't allocated by Go at all.

Because data pages were always created using `NewBufferBytes` I don't believe I added a `Release` function to them. At this point I think the safest solution is to add a `Release` to the `DataPage` interface and corresponding structs, ensure it gets called by the page reader when it sets the current page to nil or otherwise, and just pass the buffer as is to the page rather than calling `NewBufferBytes`.

Reply:
Well, we can't simply pass the buffer as is to the page because the buffer itself may or may not be compressed. If the page is compressed, this function decompresses the page, and creates a new buffer with the decompressed data. Or are you proposing that we overwrite `buf` with the decompressed data in that case?

Also, if we pass the buffer as is, is there a risk that `io.ReadFull(p.r, p.buf.Bytes())` will modify the data for the 'previous' page? Or should we do what I do now and create a new buffer for each call to `Next` (as we were doing previously but at the end of the function instead).

Reply:
Ok, looking back at the code I notice something: In all cases *except* for a V2 DataPage, even if the data isn't compressed, we end up copying it out of `buf` into memory allocated by the `decompressBuffer`. This has two effects:

1. It separates the connection between `p.buf` and the `data` variable we pass to the data page.
2. It *ensures* that we can't possibly have a situation of leaking memory because we're not using the passed in memory allocator to create that copy. We know for a fact that Go allocated that memory for the page (Not necessarily a good thing, but sure, we can go with that for now. a larger refactor may be a future idea to improve this). The only time this isn't the case is line 586, where we do `io.ReadFull(p.r p.buf.Bytes())`. 

My original idea here was that we'd utilize the same buffer where possible, creating a new buffer for `p.buf` each time. This is exemplified by the `Release` that exists on the `Page` interface. So the easy solution I can see is to keep your change to create a New buffer for each call into Next, and make a slight modification in the `DataPageV2` case:

Currently it looks like this in your change.

```go
var data []byte
if compressed {
        if levelsBytelen > 0 {
                io.ReadFull(p.r, buf.Bytes()[:levelsBytelen])
        }
	if data, p.err = p.decompress(lenCompressed-levelsBytelen, buf.Bytes()[levelsBytelen:]); p.err != nil {
	        return false
        }
} else {
        io.ReadFull(p.r, buf.Bytes())
	data = buf.Bytes()
}
```

And then we call `NewBufferBytes(data)` when creating the page. We can likely solve the issue by instead doing this:

```go
var pagebuf *memory.Buffer
if compressed {
        if levelsBytelen > 0 {
                io.ReadFull(p.r, buf.Bytes()[:levelsBytelen])
        }
        var data []byte
	if data, p.err = p.decompress(lenCompressed-levelsBytelen, buf.Bytes()[levelsBytelen:]); p.err != nil {
	        return false
        }
        pagebuf = memory.NewBufferBytes(data)
} else {
        io.ReadFull(p.r, buf.Bytes())
        pagebuf = buf
        pagebuf.Retain()
}
```

And then use pagebuf to construct the `DataPageV2`. That should be sufficient, it'll still make a new buffer for each call to `Next` and allow the defer to release the memory, but it allows us to avoid the copy in this case by just passing the buffer to the Page safely.

Does all of that make sense? I kind of rambled a bit but I hope it all made sense and came across well.

Reply:
Yes, that makes sense. PR updated.

I just wonder if `memory.NewBufferBytes(buf.Bytes())` would end up copying the `[]byte` or not? Since both `memory.NewBufferBytes` and `buf.Bytes` both use/pass references to the underlying `[]byte` and does not do any `copy`'s

Reply:
`NewBufferBytes(buf.Bytes())` just creates a `*memory.Buffer` which references the existing slice, it doesn't copy the contents of the slice. So there's no copying happening.